File: Java-Core/src/main/java/presto/type/CollectionType.java
package presto.type;

import presto.error.SyntaxError;
import presto.runtime.Context;

public abstract class CollectionType extends NativeType {

	IType itemType;
	
	protected CollectionType(String name, IType itemType) {
		super(name);
		this.itemType = itemType;
	}
	
	public IType getItemType() {
		return itemType;
	}
	
	@Override
	public void checkExists(Context context) throws SyntaxError {
		itemType.checkExists(context);
	}
	
	@Override
	public IType checkContains(Context context, IType other) throws SyntaxError {
		if(itemType.isAssignableTo(context, other))
			return BooleanType.instance();
		else
			return super.checkContains(context, other);
	}
	
}

File: Java-CodeServer/src/main/resources/js/prompto/type/CollectionType.js
var NativeType = require("./NativeType").NativeType;
var BooleanType = require("./BooleanType").BooleanType;

function CollectionType(name, itemType) {
	NativeType.call(this, name);
	this.itemType = itemType;
	return this;
}

CollectionType.prototype = Object.create(NativeType.prototype);
CollectionType.prototype.constructor = CollectionType;

/*
public IType getItemType() {
	return itemType;
}

@Override
public void checkExists(Context context) throws SyntaxError {
	itemType.checkExists(context);
}
*/

CollectionType.prototype.checkContains = function(context, other) {
	if(this.itemType.isAssignableTo(context, other)) {
		return BooleanType.instance;
	} else {
		return NativeType.prototype.checkContains.call(this, context, other);
	}
};

exports.CollectionType = CollectionType;

